src/pkcs11/framework-pkcs15.c

```diff
@@ -1207,7 +1207,8 @@ pkcs15_init_slot(struct sc_pkcs15_card *p15card, struct sc_pkcs11_slot *slot,
 						p15card->tokeninfo ? p15card->tokeninfo->label : "",
 						32);
 			}
-			slot->token_info.flags |= CKF_LOGIN_REQUIRED;
+			if (p15card->tokeninfo->flags & SC_PKCS15_TOKEN_LOGIN_REQUIRED)
```

Comment:
I don't know. But assuming that the card needs a logon if PIN object is present is at least in this case wrong (especially since the card says otherwise). And I think if the card erroneously reports that it doesn't need a logon, then it's for the card driver to fix this?

Reply:
Well, at least in the case of eOI low-level/pin-less app there are no cryptographic functions that a user MUST be logged in to perform, so omitting the `CKF_LOGIN_REQUIRED` is in place.

Perhaps other drivers should set `SC_PKCS15_TOKEN_LOGIN_REQUIRED` (why is there a `Don't use` comment)? 

Reverting this commit would break eOI pin-less operation, so I would prefer not to do it?

Reply:
the "Don't use" comment comes from a time where the flag actually was used... 50f99a20f8. ISO 7816-15 actually has a card flag, which seems to have been translated in this flag:
![grafik](https://user-images.githubusercontent.com/1266577/234428105-f7f025a2-efee-44e1-8c15-f26543b658dd.png)
However, that flag is given without explanation and doesn't actually make sense to reflect the selective (pin/non-pin) behavior you are looking for, @llogar .

First, I suggest we revert this commit.
Second, for eOI, you want to have a PKCS#11 slot that is *not* associated with a PIN as required authentication object. Hence, you want to land here:
https://github.com/OpenSC/OpenSC/blob/aedf3e7acfbd5e4acfc25d9392cf5d7fda033b29/src/pkcs11/framework-pkcs15.c#L1705-L1709
And next you want your objects that are not bound to a PIN be added here:
https://github.com/OpenSC/OpenSC/blob/aedf3e7acfbd5e4acfc25d9392cf5d7fda033b29/src/pkcs11/framework-pkcs15.c#L1711-L1714
If the public data for eOID is not added here, please have a look at how `_add_public_objects()` finds them.

Reply:
Although private key in the low level app is not protected by a PIN and ECDSA (the only cryptographic function the card can do) can be done without authenticating first, a non-empty AODF is still present, as it seems that PINs are shared between both on-card apps (the one that doesn't use them and the one that does), so one can manipulate (change/unblock) PINs when either of them is selected.

So I would still prefer to have some sort of flag to set/unset `CKF_LOGIN_REQUIRED` if required (or have it set by default, and unset it if required only). If `SC_PKCS15_TOKEN_LOGIN_REQUIRED` is not suitable for this (and I can't quite understand your reasoning), then perhaps introduce a new one, maybe in p15card->opts.login_not_required or perhaps add `SC_PKCS15_CARD_FLAG_LOGIN_NOT_REQUIRED` to p15card->flags?

Reply:
If you have some key that does *not* need a PIN to be verified, then don't set the `auth_id` object in that `struct sc_pkcs15_object`.

Maybe I am undertanding too little from what you actually want to achieve. Could you paste your (current) PKCS#15 profile (`pkcs15-tool -D`) and state which of those keys are do not need a PIN?

Reply:
Here's the output of the `pkcs15-tool -D` and as I've mentioned previously, no login is required for any and all card operations when using the pin-less app (Prijava brez PIN-a).
```
Using reader with a card: Gemalto Ezio Shield (S15A0238574896) 00 00
PKCS#15 Card [Prijava brez PIN-a]:
	Version        : 1
	Serial number  : ********************
	Manufacturer ID: NXP
	Last update    : 20220401907405
	Flags          : Read-only, PRN generation

PIN [Norm PIN]
	Object Flags   : [0x00]
	Auth ID        : 4e6f726d2050554b
	ID             : 4e6f726d2050494e
	Flags          : [0x1D], case-sensitive, change-disabled, unblock-disabled, initialized
	Length         : min_len:6, max_len:6, stored_len:6
	Pad char       : 0x00
	Reference      : 5 (0x05)
	Type           : UTF-8
	Path           : 0005

PIN [Norm PUK]
	Object Flags   : [0x00]
	ID             : 4e6f726d2050554b
	Flags          : [0x59], case-sensitive, unblock-disabled, initialized, unblockingPin
	Length         : min_len:10, max_len:10, stored_len:10
	Pad char       : 0x00
	Reference      : 6 (0x06)
	Type           : UTF-8
	Path           : 0006

Private EC Key [{bfeeb6fb-1677-f82b-5134-992d9ddcbcb5}]
	Object Flags   : [0x00]
	Usage          : [0x104], sign, derive
	Access Flags   : [0x01], sensitive
	Algo_refs      : 0
	FieldLength    : 384
	Key ref        : 2 (0x02)
	Native         : yes
	Path           : e828bd080f014e585031::0021
	ID             : 7b62666565623666622d313637372d663832622d353133342d3939326439646463626362357d
	MD:guid        : bcb7a2ad-cc10-2ed8-5ef0-958b58b5efa1

X.509 Certificate [{bfeeb6fb-1677-f82b-5134-992d9ddcbcb5}]
	Object Flags   : [0x02], modifiable
	Authority      : no
	Path           : e828bd080f014e585031::001d
	ID             : 7b62666565623666622d313637372d663832622d353133342d3939326439646463626362357d
	Encoded serial : 02 14 468C9EE761D939FD35B38F0D7C5A679C6E604239

X.509 Certificate [{72664987-f3cd-4d8e-bddb-6ca3da786ef4}]
	Object Flags   : [0x00]
	Authority      : yes
	Path           : e828bd080f014e585031::001c
	ID             : 7b37323636343938372d663363642d346438652d626464622d3663613364613738366566347d
	Encoded serial : 02 14 7605145C85CCD2CCF901CDAA55DC11AD18303272

X.509 Certificate [{3ddda0ab-3cb8-4d33-90ab-079c67be98ac}]
	Object Flags   : [0x00]
	Authority      : yes
	Path           : e828bd080f014e585031::001b
	ID             : 7b33646464613061622d336362382d346433332d393061622d3037396336376265393861637d
	Encoded serial : 02 14 5E851CBCFD5CDA6D08FDE17FBB84EE0F966E62F1

Data object <0>
	applicationName: 
	applicationOID:  1.3.6.1.4.1.28137.61166
	Path:            e828bd080f014e585031::dc00
	Data (336 bytes): ******

Data object <1>
	applicationName: 
	applicationOID:  1.3.6.1.4.1.28137.1000
	Path:            e828bd080f014e585031::dc02
	Data (18 bytes): ************************************

```
And here is the output of the `pkcs11-tool --list-slots` with the "official" PKCS#11 dll (one slot per each on-card app):
```
C:\> pkcs11-tool.exe --module C:\windows\system32\asepkcs.dll --list-slots
Available slots:
Slot 0 (0x0): Gemalto IDBridge CT7xx 0
  token label        : Prijava brez PIN-a
  token manufacturer : NXP
  token model        : ChipDocLite
  token flags        : rng, token initialized, PIN initialized, readonly
  hardware version   : 1.0
  firmware version   : 1.0
  serial num         : ****************
  pin min/max        : 6/6
Slot 1 (0x1): Gemalto IDBridge CT7xx 0#1
  token label        : Podpis in prijava
  token manufacturer : NXP
  token model        : ChipDocLite
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 1.0
  firmware version   : 1.0
  serial num         : ****************
  pin min/max        : 6/6

```
I would like to achieve the same output (of course, no separate slots for separate on-card apps). The code in question achieved this, both for `pkcs11-tool --list-slots` (and `pkcs11-tool --list-objects`):
```
$ pkcs11-tool --list-slots
Available slots:
Slot 0 (0x0): Gemalto Ezio Shield (S15A0238574896) 00 00
  token label        : Prijava brez PIN-a (Norm PIN)
  token manufacturer : NXP
  token model        : ChipDocLite
  token flags        : token initialized, PIN initialized, readonly
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : ****************
  pin min/max        : 6/6

```
My, not so elegant, solution to the problem would be something like this commit: https://github.com/llogar/OpenSC/commit/8b56248bb5e5a79fcd5337e12e630528bf34b180
Here the `CKF_LOGIN_REQUIRED` can be unset on a per card driver basis and the default behaviour is kept unchanged and compatibility with skeid retained?

Reply:
I reverted the commit in question, just now.

From your PKCS#15 profile above, `Private EC Key [{bfeeb6fb-1677-f82b-5134-992d9ddcbcb5}]` does not have an `Auth ID` associated, which means that it doesn't require a PIN validation before use. This seems to be what you want and the private key should be registered as public object in [a second seperate slot which does *not* set the `CKF_LOGIN_REQUIRED` flag](https://github.com/OpenSC/OpenSC/blob/master/src/pkcs11/framework-pkcs15.c#L1713). Could you please check why this is not the case for you?

Reply:
Or, I could (if login is not required, i.e. `SC_PKCS15_TOKEN_LOGIN_REQUIRED` is not set in tokeninfo) just delete all the authentication objects in `sc_pkcs15emu_eoi_init_ex()`. In this case it wouldn't be possible to change/unblock PIN when pin-less app is used, but the outputs of `pkcs15-tool -D` and `pkcs11-tool --list-slots` would look fine:
```
PKCS#15 Card [Prijava brez PIN-a]:
	Version        : 1
	Serial number  : ********************
	Manufacturer ID: NXP
	Last update    : 20220401907405
	Flags          : Read-only, PRN generation

Private EC Key [{bfeeb6fb-1677-f82b-5134-992d9ddcbcb5}]
	Object Flags   : [0x00]
	Usage          : [0x104], sign, derive
	Access Flags   : [0x01], sensitive
	Algo_refs      : 0
	FieldLength    : 384
	Key ref        : 2 (0x02)
	Native         : yes
	Path           : e828bd080f014e585031::0021
	ID             : 7b62666565623666622d313637372d663832622d353133342d3939326439646463626362357d
	MD:guid        : bcb7a2ad-cc10-2ed8-5ef0-958b58b5efa1

X.509 Certificate [{bfeeb6fb-1677-f82b-5134-992d9ddcbcb5}]
	Object Flags   : [0x02], modifiable
	Authority      : no
	Path           : e828bd080f014e585031::001d
	ID             : 7b62666565623666622d313637372d663832622d353133342d3939326439646463626362357d
	Encoded serial : 02 14 468C9EE761D939FD35B38F0D7C5A679C6E604239

X.509 Certificate [{72664987-f3cd-4d8e-bddb-6ca3da786ef4}]
	Object Flags   : [0x00]
	Authority      : yes
	Path           : e828bd080f014e585031::001c
	ID             : 7b37323636343938372d663363642d346438652d626464622d3663613364613738366566347d
	Encoded serial : 02 14 7605145C85CCD2CCF901CDAA55DC11AD18303272

X.509 Certificate [{3ddda0ab-3cb8-4d33-90ab-079c67be98ac}]
	Object Flags   : [0x00]
	Authority      : yes
	Path           : e828bd080f014e585031::001b
	ID             : 7b33646464613061622d336362382d346433332d393061622d3037396336376265393861637d
	Encoded serial : 02 14 5E851CBCFD5CDA6D08FDE17FBB84EE0F966E62F1

Data object <0>
	applicationName: 
	applicationOID:  1.3.6.1.4.1.28137.61166
	Path:            e828bd080f014e585031::dc00
	Data (336 bytes): ******

Data object <1>
	applicationName: 
	applicationOID:  1.3.6.1.4.1.28137.1000
	Path:            e828bd080f014e585031::dc02
	Data (18 bytes): ************************************

```
and
```
Slot 0 (0x0): Gemalto Ezio Shield (S15A0238574896) 00 00
  token label        : Prijava brez PIN-a
  token manufacturer : NXP
  token model        : ChipDocLite
  token flags        : token initialized, readonly
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : ****************
  pin min/max        : 4/8

```
The only difference is missing `PIN initialized` in token flags, but I don't think it matters...

The needed change is here: https://github.com/llogar/OpenSC/commit/51a093f446f4f1c43660ec0da9f3f3fa9019192d

Which way to go?

Reply:
That's not what I mean. Looking at your [initial profile](https://github.com/OpenSC/OpenSC/pull/2646#discussion_r1181696419), OpenSC *should* actually create two slots: One slot for PIN one for the public objects without authentication, i.e. for your private key. Please set a break point at https://github.com/OpenSC/OpenSC/blob/master/src/pkcs11/framework-pkcs15.c#L1713 and check whether your public key gets registered in the new slot.

Reply:
> From your PKCS#15 profile above, `Private EC Key [{bfeeb6fb-1677-f82b-5134-992d9ddcbcb5}]` does not have an `Auth ID` associated, which means that it doesn't require a PIN validation before use. This seems to be what you want and the private key should be registered as public object in [a second seperate slot which does _not_ set the `CKF_LOGIN_REQUIRED` flag](https://github.com/OpenSC/OpenSC/blob/master/src/pkcs11/framework-pkcs15.c#L1713). Could you please check why this is not the case for you?

Well, since a PIN object exists, a slot is created (although without any objects) and thus the !slot (in your code excerpt) is false and your code never executes, the public objects are added in the next step (to the previously created slot)...

Reply:
Thank you for clearifying. since I don't have such a token, I didn't know what exactly what was happening.

You're right that the current logic does *not* create a second slot for all the public objects (I was previously wrong about this). Instead, all public objects are added to the PIN-slot (even though they don't require a PIN validation):
https://github.com/OpenSC/OpenSC/blob/7b782d5ce3ec1b1d78251332ab391122b6da0cc6/src/pkcs11/framework-pkcs15.c#L1716-L1719

Setting `CKF_LOGIN_REQUIRED` to this single slot, means that ["there are *some* cryptographic functions that a user MUST be logged in to perform"](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html). So a good application would still check the authId to check whether or not a PIN validation is *required* in order to use the private key.

Since I know that changing the workflow of your application is most likely not possible, you could try to single out the PIN objects from the rest of the objects. You already did that by removing the PIN objects above, which gives a slot that doesn't have `CKF_LOGIN_REQUIRED` set. Now, instead of removing them, you can attach the PINs to an additional virtual card application [in `sc_card_t`](https://github.com/OpenSC/OpenSC/blob/7b782d5ce3ec1b1d78251332ab391122b6da0cc6/src/libopensc/opensc.h#L592-L593), which [should then be bound in a separate slot](https://github.com/OpenSC/OpenSC/blob/7b782d5ce3ec1b1d78251332ab391122b6da0cc6/src/pkcs11/slot.c#L347-L373). That would involve creating a virtual AID (`sc_aid`), which can be used by `sc_pkcs15emu_eoi_init_ex()` to check whether *only* the PIN objects should be propagated or whether everything *but* the PIN objects should be propagated to the PKCS#15 profile.

Note that we already have some workarounds in our code built in for PKCS#11 applications, that aren't very nice (NSS :eyes:). If you clearify on your application/use case we may consider adding more workarounds. However, fixing the application would make sense, imho.

Reply:
I don’t have (or work on) a specific application in mind. I think the only applications that most probably use pin-less applet of our eID card are applications for health insurance/medical professionals, where eID card can be used instead of the old health insurance card. 
And I think all this software runs on windows and uses the official middleware, anyway.

But I still have an issue with the `CKF_LOGIN_REQUIRED` flag being set, as it is now:

a) When eOI pin-less applet is being used, there are clearly **NO** cryptographic functions that a user MUST be logged in to perform, so `CKF_LOGIN_REQUIRED` should not be set even if there are PIN objects on the card. The output of the `pkcs11-tool –list-slots` (with official dll, which is my reference here)
```
Slot 0 (0x0): Gemalto IDBridge CT7xx 0
  token label        : Prijava brez PIN-a
  token manufacturer : NXP
  token model        : ChipDocLite
  token flags        : rng, token initialized, PIN initialized, readonly
  hardware version   : 1.0
  firmware version   : 1.0
  serial num         : ****************
  pin min/max        : 6/6
```
shows that `CKF_LOGIN_REQUIRED` is not set, but `CKF_USER_PIN_INITIALIZED` is, which, I guess, tells the calling application that PIN/authentication objects are present and authentication should be used to perform some non-cryptographic functions (change PIN, perhaps alter some data)?

b) _loginRequired_ in PKCS#15 _tokenFlags_ is explained as "_if login (i.e., authentication) is required before accessing any data_" and I think one could easily translate this to PKCS#11 `CKF_LOGIN_REQUIRED`  "_True if there are some cryptographic functions that a user MUST be logged in to perform_" and should be treated similarly. Or at the least have a way to handle it from a card driver and not just assume that if there are PIN/authentication objects, `CKF_LOGIN_REQUIRED` should be set, as it is now.

c) and finally, during development and testing of eOI support, I have noticed that if `CKF_LOGIN_REQUIRED` flag is set some applications (although I can't remember which ones, maybe even Firefox), want to perform a login (and ask for PIN) in order to search for private objects (keys and/or certificates). And since, I think, all populated slots are processed, the solution with an extra slot that has `CKF_LOGIN_REQUIRED` would popup a login window, which is not what would one want…

In any case, if there won't be any change how `CKF_LOGIN_REQUIRED` is handled in the common code, I think I'll just go with deleting the PIN objects for pin-less slot, as I've done in my 2nd example/commit. I think it's just not worth the effort to implement a separate slot for PIN management (and also because of issue c) from above), as this can easily be done when QES applet/slot is used, which is probably a more common usage scenario anyway…

Reply:
OK, let me know when you have some code ready that deletes the PIN object. I'll try to add something on top in order to push the PIN objects into a different card application (i.e. different PKCS#11 slot).